crates/subspace-farmer/src/farmer_cache.rs

```diff
@@ -176,38 +190,38 @@ where
             // TODO: Consider implementing optional re-sync of the piece instead of just forgetting
             WorkerCommand::ForgetKey { key } => {
                 let mut caches = self.piece_caches.write().await;
+                let Some(offset) = caches.stored_pieces.remove(&key) else {
+                    // key not exist.
+                    return;
+                };
+                let cache_index = usize::from(offset.cache_index);
+                let Some(backend) = caches.backends.get(cache_index).cloned() else {
+                    // cache not exist.
+                    return;
+                };
 
-                for (cache_index, cache) in caches.iter_mut().enumerate() {
-                    let Some(offset) = cache.stored_pieces.remove(&key) else {
-                        // Not this disk farm
-                        continue;
-                    };
-
-                    // Making offset as unoccupied and remove corresponding key from heap
-                    cache.free_offsets.push_front(offset);
-                    match cache.backend.read_piece_index(offset).await {
-                        Ok(Some(piece_index)) => {
-                            worker_state.heap.remove(KeyWrapper(piece_index));
-                        }
-                        Ok(None) => {
-                            warn!(
-                                %cache_index,
-                                %offset,
-                                "Piece index out of range, this is likely an implementation bug, \
-                                not freeing heap element"
-                            );
-                        }
-                        Err(error) => {
-                            error!(
-                                %error,
-                                %cache_index,
-                                ?key,
-                                %offset,
-                                "Error while reading piece from cache, might be a disk corruption"
-                            );
-                        }
+                caches.free_offsets.push_front(offset);
+                match backend.read_piece_index(offset.piece_offset).await {
+                    Ok(Some(piece_index)) => {
+                        worker_state.heap.remove(KeyWrapper(piece_index));
+                    }
+                    Ok(None) => {
+                        warn!(
+                            %cache_index,
+                            cache_offset = %offset.piece_offset,
```

Comment:
I'd extract `piece_offset` into a variable like `cache_index` and call it `piece_offset` here as well. It is harder to debug when the same thing is called `piece_offset` in one place and `cache_offset` in another.